File: R/stats-feature.R
# Approach A: per-metabolite group testing (DESIGN.md §6.1).

# limma moderated t. Returns a data.frame ordered to match rownames(assay_log).
.test_limma <- function(assay_log, group, padjust) {
  group <- droplevels(as.factor(group))
  if (length(levels(group)) != 2L) {
    stop("limma test requires exactly two non-empty group levels; got ",
         length(levels(group)), ".", call. = FALSE)
  }
  design <- stats::model.matrix(~ group)
  fit <- limma::lmFit(assay_log, design)
  fit <- limma::eBayes(fit)
  tt <- limma::topTable(fit, coef = 2L, number = Inf, sort.by = "none")
  out <- data.frame(
    metabolite = rownames(tt),
    log2FC     = tt$logFC,
    AveExpr    = tt$AveExpr,
    t          = tt$t,
    P.Value    = tt$P.Value,
    stringsAsFactors = FALSE
  )
  for (m in padjust) {
    out[[paste0("padj_", m)]] <- stats::p.adjust(out$P.Value, method = m)
  }
  out
}

# Wilcoxon rank-sum, two groups only. Per-metabolite log2FC = mean(g2) -
# mean(g1) on the log-transformed values.
.test_wilcoxon <- function(assay_log, group, padjust) {
  group <- droplevels(as.factor(group))
  levels_g <- levels(group)
  if (length(levels_g) != 2L) {
    stop("Wilcoxon test requires exactly two non-empty group levels; got ",
         length(levels_g), ".", call. = FALSE)
  }
  g1 <- group == levels_g[1L]
  g2 <- group == levels_g[2L]
  n <- nrow(assay_log)
  pvals  <- rep(NA_real_, n)
  fcs    <- rep(NA_real_, n)
  for (i in seq_len(n)) {
    x1 <- assay_log[i, g1]; x1f <- x1[is.finite(x1)]
    x2 <- assay_log[i, g2]; x2f <- x2[is.finite(x2)]
    if (length(x1f) < 1L || length(x2f) < 1L) next
    res <- tryCatch(
      suppressWarnings(stats::wilcox.test(x2f, x1f, exact = FALSE)),
      error = function(e) NULL
    )
    pvals[i] <- if (is.null(res)) NA_real_ else res$p.value
    fcs[i]   <- mean(x2f) - mean(x1f)
  }
  out <- data.frame(
    metabolite = rownames(assay_log),
    log2FC     = fcs,
    P.Value    = pvals,
    stringsAsFactors = FALSE
  )
  for (m in padjust) {
    out[[paste0("padj_", m)]] <- stats::p.adjust(out$P.Value, method = m)
  }
  out
}

#' Feature-level group testing of metabolites
#'
#' Approach A from DESIGN.md §6.1: treat each metabolite as an individual
#' feature and run a group comparison via limma moderated t (default) or
#' Wilcoxon rank-sum, with multiple-testing correction. Concentrations are
#' log2-transformed with an adaptive per-feature pseudocount before testing
#' (DESIGN.md §4.5). Outlier samples (`@sample_outliers`) are excluded.
#'
#' Metabolite-to-metabolite correlation is high within a chemical class, so
#' the BH/BY pair is reported by default (decisions.md #29).
#'
#' @param prep A [PreprocessedData] object.
#' @param group Either the name of a column in `prep@sample@sample_meta` or
#'   a vector of length `nrow(prep@sample@assay)` (all samples, before
#'   outlier exclusion).
#' @param method One of `"limma"` (default) or `"wilcoxon"`.
#' @param padjust Character vector of multiple-testing methods to report;
#'   each becomes a `padj_<method>` column. Default `c("BH", "BY")`.
#' @return A data.frame with one row per metabolite.
#' @export
test_metabolites <- function(prep, group,
                             method = c("limma", "wilcoxon"),
                             padjust = c("BH", "BY")) {
  method <- match.arg(method)
  group <- .resolve_group(group, prep)
  keep <- !prep@sample_outliers
  group <- group[keep]
  assay <- prep@sample@assay[keep, , drop = FALSE]
  assay_log <- log2_with_pseudocount(t(assay))
  rownames(assay_log) <- prep@sample@metabolite_names
  if (method == "limma") {
    .test_limma(assay_log, group, padjust)
  } else {
    .test_wilcoxon(assay_log, group, padjust)
  }
}

File: python/metabosetr/stats_feature.py
"""Approach A: per-metabolite group testing (port of ``stats-feature.R``).

DESIGN.md §6.1. Each metabolite is an individual feature. Default engine is
the empirical-Bayes moderated t via InMoose (a drop-in Python port of R
limma, near-identical results); Wilcoxon rank-sum via SciPy is the support
option (MIGRATION.md §2 decision A). Concentrations are log2-transformed
with the adaptive per-feature pseudocount before testing; outlier samples
are excluded.
"""

from __future__ import annotations

import numpy as np
import pandas as pd

from ._padjust import p_adjust
from .transform import log2_with_pseudocount


def _resolve_group(group, prep):
    """Resolve ``group`` to a length-``n_samples`` array (before outlier drop).

    Accepts a single ``sample_meta`` column name or a length-n vector.
    """
    meta = prep.sample.sample_meta
    n = prep.sample.assay.shape[0]
    if isinstance(group, str):
        if group not in meta.columns:
            raise ValueError(
                f"`group` column '{group}' not in sample_meta. "
                f"Available: {', '.join(map(str, meta.columns))}"
            )
        return meta[group].to_numpy()
    group = np.asarray(group, dtype=object)
    if group.shape[0] != n:
        raise ValueError(
            "`group` must be a single sample_meta column name or a vector of "
            "length nrow(sample.assay)."
        )
    return group


def _two_levels(group):
    """Sorted, dropped-empty factor levels; error unless exactly two."""
    vals = [g for g in group if g is not None and not (isinstance(g, float) and np.isnan(g))]
    levels = sorted(set(vals))
    if len(levels) != 2:
        raise ValueError(
            f"test requires exactly two non-empty group levels; got {len(levels)}."
        )
    return levels


def _add_padjust(out: pd.DataFrame, padjust) -> pd.DataFrame:
    for m in padjust:
        out[f"padj_{m}"] = p_adjust(out["P.Value"].to_numpy(), m)
    return out


def _test_wilcoxon(assay_log, group, padjust, metabolite_names) -> pd.DataFrame:
    from scipy.stats import mannwhitneyu

    levels = _two_levels(group)
    group = np.asarray(group, dtype=object)
    g1 = group == levels[0]
    g2 = group == levels[1]
    n = assay_log.shape[0]
    pvals = np.full(n, np.nan)
    fcs = np.full(n, np.nan)
    for i in range(n):
        x1 = assay_log[i, g1]
        x2 = assay_log[i, g2]
        x1f = x1[np.isfinite(x1)]
        x2f = x2[np.isfinite(x2)]
        if x1f.size < 1 or x2f.size < 1:
            continue
        try:
            res = mannwhitneyu(
                x2f, x1f, alternative="two-sided", use_continuity=True, method="asymptotic"
            )
            pvals[i] = res.pvalue
        except ValueError:
            pvals[i] = np.nan
        fcs[i] = x2f.mean() - x1f.mean()
    out = pd.DataFrame(
        {"metabolite": metabolite_names, "log2FC": fcs, "P.Value": pvals}
    )
    return _add_padjust(out, padjust)


def _test_limma(assay_log, group, padjust, metabolite_names) -> pd.DataFrame:
    from inmoose.limma import eBayes, lmFit, topTable

    levels = _two_levels(group)
    group = np.asarray(group, dtype=object)
    # design: intercept + indicator for the second (reference-sorted) level,
    # matching R model.matrix(~ group) with alphabetically-sorted levels.
    g2 = (group == levels[1]).astype(float)
    design = np.column_stack([np.ones(len(group)), g2])

    expr = pd.DataFrame(assay_log, index=metabolite_names)
    fit = eBayes(lmFit(expr, design))
    # InMoose labels the design columns column0/column1; the group effect is
    # column1. sort_by="none" is unsupported, so sort then restore feature
    # order via reindex to match R topTable(sort.by="none").
    tt = topTable(fit, coef="column1", number=assay_log.shape[0], sort_by="p")
    tt = tt.reindex(metabolite_names)
    out = pd.DataFrame(
        {
            "metabolite": metabolite_names,
            "log2FC": tt["log2FoldChange"].to_numpy(),
            "AveExpr": tt["AveExpr"].to_numpy(),
            "t": tt["stat"].to_numpy(),
            "P.Value": tt["pvalue"].to_numpy(),
        }
    )
    return _add_padjust(out, padjust)


def test_metabolites(prep, group, method="limma", padjust=("BH", "BY")) -> pd.DataFrame:
    """Feature-level group testing (Approach A).

    ``method`` is ``"limma"`` (default, InMoose moderated t) or
    ``"wilcoxon"``. Returns one row per metabolite.
    """
    if method not in ("limma", "wilcoxon"):
        raise ValueError("method must be 'limma' or 'wilcoxon'.")
    group = _resolve_group(group, prep)
    keep = ~prep.sample_outliers
    group = group[keep]
    assay = prep.sample.assay[keep, :]
    assay_log = log2_with_pseudocount(assay.T)  # rows = metabolites
    names = list(prep.sample.metabolite_names)
    if method == "limma":
        return _test_limma(assay_log, group, padjust, names)
    return _test_wilcoxon(assay_log, group, padjust, names)
